File: javascript/Trees/Tree.js


module.exports= {

    
    BinarySearchTree : class BinaryTree {
        
        constructor(){
           
        }

        insert(data){
            root = insert(data , root);
        }

        print(){
            print(root);
        }

        Remove(element){
            root = Remove(element , root);
        }
    }



}

/**
 * @type {BinaryNode}
 */
var root  = null;

class BinaryNode {

    constructor(data , left , right) {
        this.data = data;
        this.left = left;
        this.right = right;
    }
}

/**
 * 
 * @param {any} data 
 * @param {BinaryNode} node 
 */
function insert(data , node){

    if(node == null){
        return new BinaryNode(data , null , null);
    }
    if(data < node.data){
        node.left = insert(data , node.left);
    }else if(data > node.data){
        node.right = insert(data , node.right);
    }else {

    }
    return node;
}

//deleting Node from the tree 

/**
 * 
 * @param {any} element 
 * @param {BinaryNode} t 
 * @returns {BinaryNode}
 */
function Remove(element , t){

    if(t === null){
        return t;
    }
    if(element < t.data){
        t.left = Remove(element , t.left);
    }else if(element > t.data){
        t.right = Remove(element , t.right);
    }else if(t.left != null && t.right !=null){
        t.data = FindMin(t.right).data;
        t.right = Remove(t.data , t.right);
    }else 
        t = (t.left != null) ? t.left : t.right;
    return t;
}

/**
 * 
 * @param {BinaryNode} t 
 */
function FindMin(t){
    if(t == null){
        return t;
    }
    while(t.left != null){
        t = t.left;
    }
    return t;
}


/**
 * 
 * @param {BinaryNode} node 
 */
function print(node){
    if(node == null){
        return;
    }
    print(node.left);
    console.log(node.data);
    print(node.right);
}

File: c#/TreesDataStructure/Tree.cs
namespace TreesDataStructure;

//coding Tree Data Structure . 
    // Binary Tree , Not Balanced Tree , worst Case O(N) Search if the insertion is Targetting one side of the Tree 


public class Tree <T> where T : IComparable<T>
{
    public Tree()
    {
        root = null; // on initializing the tree the root will be null
    }

    public void insert(T element)
    {
        root = insert(element,root!);
    }

    ///deleting Nodes : 
    //case 1 : the target Node has no children . [easy to delete]
    //case 2 : the target Node has one child -> right one or left one 
    //case 3 : the target node has two children [worst one]
        //handling the 3rd case : simply replace the value with the smallest right left Node value 
        //then remove the duplicated tree .
    public void Remove(T element){
        root = Remove(element , root!);
    }   

    private Node? Remove(T element, Node t){
        if(t == null){
            return t;
        }
        if(element.CompareTo(t.data) < 0){
            t.left = Remove(element,t.left);
        }else if(element.CompareTo(t.data) > 0){
            t.right = Remove(element,t.right);
        }else if(t.left != null && t.right != null){
            t.data = FindMin(t.right)!.data;
            t.right = Remove(t.data , t.right); // remove the item
        }else {
            t = (t.left != null) ? t.left : t.right!;
        }
        return t;
    }

    public T? FindMin(){
        return FindMin(root!)!.data ?? default(T)!;
    }      

    private Node? FindMin(Node t){

        if(t == null){
            return null;
        }

        while(t.left != null){
            t = t.left;
        }

        return t;

    }
    private Node insert(T element, Node t)
    {
        if (t == null)
        {
            return new Node(element, null, null);
        }

        if (element.CompareTo(t.data) < 0)
        {
            t.left = insert(element , t.left);
            
        }else if (element.CompareTo(t.data) > 0)
        {
            t.right = insert(element , t.right);
        }
        else
        {
            //duplicated entry don't do anything 
        }

        return t;
    }


    public void Print()
    {
        Print(root!);
    }

    private void Print(Node t)
    {
        if (t == null)
        {
            return;
        }
        
        Print(t.left);
        Console.WriteLine(t.data);
        Print(t.right);
    }
    
    
    private class Node
    {
        public Node(T element , Node? l = null , Node? r = null)
        {
            data = element;
            left = l;
            right = r;
        }
        public Node(T element) : this(element , null , null)
        {
            
        }
        
        public  T data;
        public Node left;
        public Node right;
    }
    
    
    
    
    
    
    
    //data
    private Node? root;

}